### PR TITLE
Service Registration: Cleanup of service.js

### LINKED ISSUE #4986: Service registration should not create the service
Currently our `installFooService` methods all make an instance of the service. This was different originally, before we started to centrally install the services in their respective JS binary.

We should switch the registrations to be just that, so that we don't make instances until we have to.

### PR DESCRIPTION
**Pull request by kmh287**
Now that no services use the `fromClass` and `fromClassForDoc` codepaths, service.js can be cleaned up a bit. This PR:

- removes `fromClass` and `fromClassForDoc`. The same functionality is provided by `registerService` + `getService` and `registerServiceForDoc` + `getServiceForDoc`
- Removes `getExistingService{,forDoc}` methods whose functionality is provided by asserts elsewhere. 
- `getExistingServiceForWindowOrNull` has been renamed to be consistent with other function names
- Removes legacy code path related to the removed functions

/to @choumx @dvoytenko 

Closes #4986

### CODE REVIEW

File: src/service.js [hunk 1]
```diff
@@ -205,65 +146,58 @@ function getLocalExistingServiceForEmbedWinOrNull(embedWin, id) {
...
- * @param {!Window} win
- * @param {string} id of the service.
- * @param {function(new:T, !Window)} constructor
- * @return {T}
- * @template T
- */
-export function fromClass(win, id, constructor) {
-  win = getTopWindow(win);
-  return getServiceInternal(win, win, id, constructor);
+  if (!isServiceRegistered(win, id)) {
```
Comment (choumx) [inline src/service.js]: The existence of `opt_factory` here and in `getServiceForDoc` breaks the registration pattern. What do you think we should do about this?

File: src/service.js [hunk 2]
```diff
@@ -363,6 +255,63 @@ export function getServicePromiseOrNullForDoc(nodeOrDoc, id) {
       id);
 }
 
+/**
+ * @param {!Window} embedWin
+ * @param {string} id
+ * @return {?Object}
+ */
+function getExistingServiceForEmbedWinOrNull(embedWin, id) {
```
Comment (choumx) [inline src/service.js]: Can you undo the change in position for these methods? It makes the diff harder to read. 

In general, recommend minimizing diffs in sensitive changes like this PR.

File: src/service.js [hunk 3]
```diff
@@ -533,10 +435,38 @@ function registerServiceInternal(holder, context, id, opt_ctor, opt_factory) {
...
+/**
+ * Get service `id` from `holder`. Assumes the service
+ * has already been registered.
+ * @param {!Object} holder Object holding the service instance.
+ * @param {!Window|!./service/ampdoc-impl.AmpDoc} context Win or AmpDoc.
+ * @param {string} id of the service.
+ * @return {Object}
+ * @template T
+ */
+function getServiceInternal(holder, context, id) {
```
Comment (choumx) [inline src/service.js]: Ditto re: position.

File: src/service.js (hunk 2 above)
Comment (kmh287) [inline src/service.js]: done

File: src/service.js (hunk 3 above)
Comment (kmh287) [inline src/service.js]: ''

File: src/service.js (hunk 1 above)
Comment (kmh287) [inline src/service.js]: The current behavior is far from ideal: if you call getService with a factory but the service is already registered then the factory is ignored. 

I can look into removing them, but I would rather not add that to this PR.

File: src/service.js [hunk 4]
```diff
@@ -550,37 +476,28 @@ function getServicePromiseInternal(holder, id) {
...
-  let s = services[id];
-  if (s) {
-    // Service is registered, but not yet instantiated.
-    s.promise = p;
-    s.resolve = resolve;
-    // Instantiate service immediately.
-    if (s.build) {
-      s.obj = s.build();
-      s.resolve(s.obj);
-    }
```
Comment (choumx) [inline src/service.js]: Removed because this is already done in `getServicePromiseOrNullInternal`?

File: src/service.js [hunk 5]
```diff
@@ -164,30 +129,35 @@ export function installServiceInEmbedScope(embedWin, id, service) {
...
 }
 
-
 /**
  * @param {!Window} embedWin
  * @param {string} id
  * @return {?Object}
  */
-function getLocalExistingServiceForEmbedWinOrNull(embedWin, id) {
+function getExistingServiceForEmbedWinOrNull(embedWin, id) {
```
Comment (choumx) [inline src/service.js]: Why remove the "Local" qualifier? Seems important vis-a-vis `getExistingServiceInEmbedScope` since the former falls back to the top-level window.

File: src/service.js [hunk 6]
```diff
@@ -307,34 +269,19 @@ export function getServicePromiseOrNull(win, id) {
...
- */
-export function fromClassForDoc(nodeOrDoc, id, constructor) {
-  const ampdoc = getAmpdoc(nodeOrDoc);
-  return getServiceInternal(
-      getAmpdocServiceHolder(ampdoc),
-      ampdoc,
-      id,
-      constructor);
+  const holder = getAmpdocServiceHolder(ampdoc);
+  if (!isServiceRegistered(holder, id)) {
```
Comment (choumx) [inline src/service.js]: Update the documentation to state that the service will be registered with `opt_factory` if unregistered.

File: src/service.js [hunk 7]
```diff
@@ -452,47 +398,28 @@ function getAmpdocService(win) {
...
-          ? new opt_constructor(context)
-          : opt_factory(context);
-    }
+    dev().assert(s.build, `Service ${id} registered without builder nor impl.`);
+    s.obj = s.build();
     // The service may have been requested already, in which case we have a
     // pending promise we need to fulfill.
-    if (s.resolve) {
-      s.resolve(s.obj);
+    if (s.promise && s.resolve) {
```
Comment (choumx) [inline src/service.js]: Is there an instance where `s.resolve` is not null but `s.promise` is?

File: src/service.js (hunk 1 above)
Comment (choumx) [inline src/service.js]: > if you call getService with a factory but the service is already registered then the factory is ignored.

No different from trying to register a service twice, right? 

> I would rather not add that to this PR.

That's fine, just thinking about how to make this API simpler.

**Comment by choumx** [comment]
Thanks for reverting the positioning changes. Diff is much more readable now.

File: src/service.js [hunk 8]
```diff
@@ -166,10 +131,16 @@ export function installServiceInEmbedScope(embedWin, id, service) {
...
       'Service override has already been installed: %s', id);
-  getServiceInternal(embedWin, embedWin, id, undefined, () => service);
+  registerServiceInternal(
+      embedWin,
+      embedWin,
+      id,
+      /* opt_ctor */ undefined,
+      () => service);
+  // Force service to build
+  getServiceInternal(embedWin, embedWin, id);
```
Comment (choumx) [inline src/service.js]: This can be reverted. Let's address the future of the convenience builder separately.

File: src/service.js [hunk 9]
```diff
@@ -452,47 +399,28 @@ function getAmpdocService(win) {
...
-    }
+    dev().assert(s.build, `Service ${id} registered without builder nor impl.`);
+    s.obj = s.build();
     // The service may have been requested already, in which case we have a
     // pending promise we need to fulfill.
     if (s.resolve) {
-      s.resolve(s.obj);
+      s.resolve(/** @type {!Object} */(s.obj));
     }
+    s.build = null;
```
Comment (choumx) [inline src/service.js]: Please add a unit test for this.

File: src/service.js [hunk 10]
```diff
@@ -532,14 +460,13 @@ function registerServiceInternal(holder, context, id, opt_ctor, opt_factory) {
...
   // The service may have been requested already, in which case there is a
   // pending promise that needs to fulfilled.
-  if (s.promise && s.resolve) {
-    const obj = s.build();
-    s.obj = obj;
-    s.resolve(obj);
+  if (s.resolve) {
+    s.obj = s.build();
+    s.resolve(/** @type {!Object} */(s.obj));
+    s.build = null;
```
Comment (choumx) [inline src/service.js]: Ditto.

File: src/service.js [hunk 11]
```diff
@@ -550,37 +477,28 @@ function getServicePromiseInternal(holder, id) {
...
+    resolve,
+    build: null,
+  };
+  return promise;
 }
 
 
 /**
+ * Returns a promise for service `id` if the service has been registered
+ * on `holder`.
```
Comment (choumx) [inline src/service.js]: Need to set `s.build` to null in this method. And a unit test. 😄

File: src/service.js [hunk 12]
```diff
@@ -452,47 +399,28 @@ function getAmpdocService(win) {
...
-          ? new opt_constructor(context)
-          : opt_factory(context);
-    }
+    dev().assert(s.build, `Service ${id} registered without builder nor impl.`);
+    s.obj = s.build();
     // The service may have been requested already, in which case we have a
     // pending promise we need to fulfill.
     if (s.resolve) {
-      s.resolve(s.obj);
+      s.resolve(/** @type {!Object} */(s.obj));
```
Comment (jridgewell) [inline src/service.js]: Can we get rid of these typecasts by marking `s.build` correctly?

File: src/service.js [hunk 13]
```diff
@@ -532,14 +460,13 @@ function registerServiceInternal(holder, context, id, opt_ctor, opt_factory) {
 
   // The service may have been requested already, in which case there is a
   // pending promise that needs to fulfilled.
-  if (s.promise && s.resolve) {
-    const obj = s.build();
-    s.obj = obj;
-    s.resolve(obj);
+  if (s.resolve) {
+    s.obj = s.build();
```
Comment (jridgewell) [inline src/service.js]: Just call `getServiceInternal`?

Comment (kmh287) [inline src/service.js]: good idea

File: src/service.js (hunk 10 above)
Comment (kmh287) [inline src/service.js]: line deleted

File: src/service.js (hunk 11 above)
Comment (kmh287) [inline src/service.js]: routing all of this through getServiceInternal to avoid duplication

File: src/service.js [hunk 14]
```diff
@@ -596,8 +512,9 @@ function getServicePromiseOrNullInternal(holder, id) {
     } else {
       dev().assert(s.build,
```
Comment (jridgewell) [inline src/service.js]: Isn't this handled by `getServiceInternal`?

File: src/service.js [hunk 15]
```diff
@@ -596,8 +512,9 @@ function getServicePromiseOrNullInternal(holder, id) {
     } else {
```
Comment (jridgewell) [inline src/service.js]: Can we collapse the previous block into this one?

File: src/service.js [hunk 16]
```diff
@@ -452,47 +399,27 @@ function getAmpdocService(win) {
...
-          ? new opt_constructor(context)
-          : opt_factory(context);
-    }
+    dev().assert(s.build, `Service ${id} registered without builder nor impl.`);
+    s.obj = s.build();
     // The service may have been requested already, in which case we have a
     // pending promise we need to fulfill.
     if (s.resolve) {
-      s.resolve(s.obj);
+      s.resolve((s.obj));
```
Comment (jridgewell) [inline src/service.js]: Nit.

Comment (kmh287) [inline src/service.js]: thanks

File: src/service.js (hunk 14 above)
Comment (kmh287) [inline src/service.js]: We can get rid of it

File: src/service.js (hunk 15 above)
Comment (kmh287) [inline src/service.js]: sure

File: src/service.js [hunk 17]
```diff
@@ -532,14 +460,12 @@ function registerServiceInternal(holder, context, id, opt_ctor, opt_factory) {
 
   // The service may have been requested already, in which case there is a
   // pending promise that needs to fulfilled.
-  if (s.promise && s.resolve) {
-    const obj = s.build();
-    s.obj = obj;
-    s.resolve(obj);
+  if (s.resolve) {
+    // getServiceInternal will resolve the project.
```
Comment (choumx) [inline src/service.js]: "resolve the promise."

File: test/functional/test-analytics.js [hunk 18]
```diff
@@ -88,8 +87,10 @@ describes.realWin('analytics', {amp: true}, env => {
       const ele = win.document.createElement('div');
       win.document.body.appendChild(ele);
       const baseEle = new BaseElement(ele);
-      fromClassForDoc(
-          ampdoc, 'amp-analytics-instrumentation', MockInstrumentation);
+      registerServiceBuilderForDoc(
+        ampdoc, 'amp-analytics-instrumentation', MockInstrumentation);
```
Comment (choumx) [inline test/functional/test-analytics.js]: Nit: +2 spaces

File: test/functional/test-service.js [hunk 19]
```diff
@@ -217,20 +201,29 @@ describe('service', () => {
...
     });
 
+    it('should set service builders to null after instantiation', () => {
+      registerServiceBuilder(window, 'a', Class);
+      expect(window.services['a'].obj).to.be.null;
+      expect(window.services['a'].build).to.not.be.null;
+      getService(window, 'a');
+      expect(window.services['a'].obj).to.not.be.null;
+      expect(window.services['a'].build).to.be.null;
+    });
```
Comment (choumx) [inline test/functional/test-service.js]: Thanks!

File: src/service.js [hunk 20]
```diff
@@ -591,13 +508,12 @@ function getServicePromiseOrNullInternal(holder, id) {
...
       return s.promise;
-    } else if (s.obj) {
-      return s.promise = Promise.resolve(s.obj);
     } else {
-      dev().assert(s.build,
-          'Expected object, promise, or builder to be present');
-      s.obj = s.build();
-      return s.promise = Promise.resolve(s.obj);
+      if (!s.obj) {
+        // Instantiate service
```
Comment (choumx) [inline src/service.js]: I think we can remove the conditional and change this comment to "Instantiate service if not already instantiated."

File: src/service.js (hunk 17 above)
Comment (kmh287) [inline src/service.js]: whoops, thanks

File: src/service.js (hunk 20 above)
Comment (kmh287) [inline src/service.js]: good idea

File: test/functional/test-analytics.js (hunk 18 above)
Comment (kmh287) [inline test/functional/test-analytics.js]: done

File: src/service.js [hunk 21]
```diff
@@ -292,49 +255,34 @@ export function getServicePromiseOrNull(win, id) {
...
 /**
  * Returns a service for the given id and ampdoc (a per-ampdoc singleton).
  * If the service is not yet available the factory function is invoked and
  * expected to return the service.
  * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
  * @param {string} id of the service.
  * @param {function(!./service/ampdoc-impl.AmpDoc):T=} opt_factory
- *     Should create the service if it does not exist yet. If the factory is
- *     not given, it is an error if the service does not exist yet.
+ *     Factory to create the service if the service does not exist yet.
```
Comment (dvoytenko) [inline src/service.js]: So, we now have `getExistingService*` for window and embed scope, but not for doc scope? Does this help to conflate registration and use-service cases/APIs?

Comment (choumx) [inline src/service.js]: I noticed this too but looks like `performanceForOrNull` uses it. If those calls can be changed to `performanceFor`, then we can remove the `getExisting...` methods altogether.

Recommend making any additional changes after this PR though.

Comment (kmh287) [inline src/service.js]: There are `getExistingService*orNull` functions, but I've completely removed functions that will fail an assertion if a service doesn't already exist as `getServiceInternal` now handles that case.

There may not be some `forDoc` ones as they weren't needed anywhere. I can add some in if you'd like. 

As for `performaneForOrNull`, my understanding is that there are times we don't expect the performance service to exist. If that's not the case then we can remove `performanceForOrNull` as @choumx suggested.

Comment (dvoytenko) [inline src/service.js]: Yeah, I think `performanceForOrNull` is used as you describe and thus it's needed. I mostly reacted to absence of `getExistingServiceForDoc` because I thought we were trying to remove API confusion between service registration and service use. E.g. Viewer is registered only once, but is being used via `viewerFor` all over the place. 

But, I'm ok if we reconsider this in a separate PR.

Comment (kmh287) [inline src/service.js]: Callers shouldn't need to worry about registered vs built. I removed those functions since they were redundant with changes made to `getServiceInternal`.

**Comment by kmh287** [comment]
@choumx @dvoytenko @jridgewell Thank you all for the very thorough reviews.
